// @sudoku/stores/gameStore.js

import { writable } from 'svelte/store';
import { decodeSencode } from '@sudoku/sencode';
import { generateSudoku } from '@sudoku/sudoku';
import { Sudoku } from '../domain/Sudoku.js';
import { Game } from '../domain/Game.js';

function emptyGrid() {
	return [
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
	];
}

function emptyState() {
	return {
		initialGrid: emptyGrid(),
		grid: emptyGrid(),
		invalidCells: [],
		won: false,
		canUndo: false,
		canRedo: false,
		currentHint: null,
		exploring: false,
		exploreFailed: false,
		canCommitExplore: false,
		canAbortExplore: false,
		ready: false,
	};
}

function createGameStore() {
	const { subscribe, set } = writable(emptyState());
	let game = null;

	function sync() {
		if (!game) {
			set(emptyState());
			return;
		}

		set({
			...game.getViewData(),
			ready: true,
		});
	}

	return {
		subscribe,

		startNew(difficulty = 'easy') {
			const puzzle = generateSudoku(difficulty);
			game = new Game(new Sudoku(puzzle));
			sync();
		},

		startCustom(source) {
			const puzzle = typeof source === 'string' ? decodeSencode(source) : source;
			game = new Game(new Sudoku(puzzle));
			sync();
		},

		guess(move) {
			if (!game) return false;
			const ok = game.guess(move);
			if (ok) sync();
			return ok;
		},

		undo() {
			if (!game) return false;
			const ok = game.undo();
			if (ok) sync();
			return ok;
		},

		redo() {
			if (!game) return false;
			const ok = game.redo();
			if (ok) sync();
			return ok;
		},

		requestCellHint(row, col) {
			if (!game) return null;
			const hint = game.getCellHint(row, col);
			sync();
			return hint;
		},

		requestNextHint() {
			if (!game) return null;
			const hint = game.getNextHint();
			sync();
			return hint;
		},

		applyCellHint(row, col) {
			if (!game) return false;
			const ok = game.applyCellHint(row, col);
			sync();
			return ok;
		},

		applyNextHint() {
			if (!game) return false;
			const ok = game.applyNextHint();
			sync();
			return ok;
		},

		clearHint() {
			if (!game) return;
			game._clearHint();
			sync();
		},

		enterExplore() {
			if (!game) return false;
			const ok = game.enterExplore();
			if (ok) sync();
			return ok;
		},

		abortExplore() {
			if (!game) return false;
			const ok = game.abortExplore();
			if (ok) sync();
			return ok;
		},

		commitExplore() {
			if (!game) return false;
			const ok = game.commitExplore();
			if (ok) sync();
			return ok;
		},

		toJSON() {
			return game ? game.toJSON() : null;
		},

		loadFromJSON(json) {
			game = Game.fromJSON(json);
			sync();
		},

		getGame() {
			return game;
		},
	};
}

export const gameStore = createGameStore();
